src/toxcore/dht/lan_discovery.rs

```diff
@@ -0,0 +1,238 @@
+    /// Create new `LanDiscovery`.
+    pub fn new(tx: Tx, dht_pk: PublicKey, ipv6: bool) -> LanDiscoverySender {
+        LanDiscoverySender {
+            tx,
+            dht_pk,
+            ipv6,
+            next_port: START_PORT,
+        }
+    }
+
+    /// Get broadcast addresses for host's network interfaces.
+    fn get_ipv4_broadcast_addrs() -> Vec<IpAddr> {
+        let ifs = get_if_addrs::get_if_addrs().expect("no network interface");
+        ifs.iter().filter_map(|interface|
+            match interface.addr {
+                IfAddr::V4(ref addr) => addr.broadcast,
+                _ => None,
+            }
+        ).map(|addr|
+            IpAddr::V4(addr)
+        ).collect()
+    }
+
+    /// Get broadcast addresses depending on IP version.
+    fn get_broadcast_addrs(&self) -> Vec<IpAddr> {
+        let mut ip_addrs = LanDiscoverySender::get_ipv4_broadcast_addrs();
+        if self.ipv6 {
+            // IPv6 broadcast address
+            ip_addrs.push("FF02::1".parse().unwrap());
+            // IPv4 global broadcast address
+            ip_addrs.push("::ffff:255.255.255.255".parse().unwrap());
+        } else {
+            // IPv4 global broadcast address
+            ip_addrs.push("255.255.255.255".parse().unwrap());
+        }
+        ip_addrs
+    }
+
+    /// Get broadcast addresses to send `LanDiscovery` packet.
+    ///
+    /// This function returns Cartesian product of addresses from
+    /// `get_broadcast_addrs` function with ports list of current iteration.
+    fn get_broadcast_socket_addrs(&mut self) -> Vec<SocketAddr> {
+        fn cycle(port: u16) -> u16 {
+            (port - START_PORT) % (END_PORT - START_PORT) + START_PORT
+        }
+        let ip_addrs = self.get_broadcast_addrs();
+        // range of ports to send discovery packet to
+        let ports_range = (self.next_port .. self.next_port + PORTS_PER_DISCOVERY).map(cycle);
+        // always send discovery packet to default port
+        let ports_range = iter::once(DEFAULT_PORT).chain(ports_range);
+        // add ports to ip addrs
+        let socket_addrs = ip_addrs.into_iter().flat_map(move |ip_addr| {
+            ports_range.clone().map(move |port| SocketAddr::new(ip_addr, port))
+        }).collect();
+        // update port for next iteration
+        self.next_port = cycle(self.next_port + PORTS_PER_DISCOVERY);
+        socket_addrs
+    }
+
+    /// Send `LanDiscovery` packets if necessary.
+    pub fn send(&mut self) -> IoFuture<()> {
+        let addrs = self.get_broadcast_socket_addrs();
+        let lan_packet = DhtPacket::LanDiscovery(LanDiscovery {
+            pk: self.dht_pk,
+        });
+
+        let stream = stream::iter_ok(
+            addrs.into_iter().map(move |addr| (lan_packet.clone(), addr))
+        );
+
+        send_all_to(&self.tx, stream)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use futures::{Future, Stream};
+
+    macro_rules! unpack {
+        ($variable:expr, $variant:path) => (
```

Comment:
How about moving this macro definition to binary_io.rs.
Because this macro is used commonly in many files.